gflownet/envs/crystals/ccrystal.py

```diff
@@ -0,0 +1,917 @@
+            raise ValueError(
+                "Detected duplicate actions between different components of Crystal "
+                "environment."
+            )
+
+        return action_space
+
+    def action2representative(self, action: Tuple) -> Tuple:
+        """
+        Replaces the continuous values of lattice parameters actions by the
+        representative action of the environment so that it can be compared against the
+        action space.
+        """
+        if self._get_stage() == Stage.LATTICE_PARAMETERS:
+            return self.subenvs[Stage.LATTICE_PARAMETERS].action2representative(
+                self._depad_action(action, Stage.LATTICE_PARAMETERS)
+            )
+        return action
+
+    def get_max_traj_length(self) -> int:
+        return sum([subenv.get_max_traj_length() for subenv in self.subenvs.values()])
+
+    def get_policy_output(self, params: dict) -> TensorType["policy_output_dim"]:
+        """
+        Defines the structure of the output of the policy model.
+
+        The policy output is in this case the concatenation of the policy outputs of
+        the three sub-environments.
+        """
+        return torch.cat(
+            [subenv.get_policy_output(params) for subenv in self.subenvs.values()]
+        )
+
+    def _get_policy_outputs_of_subenv(
+        self, policy_outputs: TensorType["n_states", "policy_output_dim"], stage: Stage
+    ):
+        """
+        Returns the columns of the policy outputs that correspond to the
+        sub-environment indicated by stage.
+
+        Args
+        ----
+        policy_outputs : tensor
+            A tensor containing a batch of policy outputs. It is assumed that all the
+            rows in the this tensor correspond to the same stage.
+
+        stage : Stage
+            Identifier of the sub-environment of which the corresponding columns of the
+            policy outputs are to be extracted.
+        """
+        init_col = 0
+        for stg, subenv in self.subenvs.items():
+            end_col = init_col + subenv.policy_output_dim
+            if stg == stage:
+                return policy_outputs[:, init_col:end_col]
+            init_col = end_col
+
+    def _get_mask_of_subenv(
+        self, mask: Union[List, TensorType["n_states", "mask_dim"]], stage: Stage
+    ):
+        """
+        Returns the columns of a tensor of masks that correspond to the sub-environment
+        indicated by stage.
+
+        Args
+        ----
+        mask : list or tensor
+            A mask of a single state as a list or a tensor containing a batch of masks.
+            It is assumed that all the rows in the this tensor correspond to the same
+            stage.
+
+        stage : Stage
+            Identifier of the sub-environment of which the corresponding columns of the
+            masks are to be extracted.
+        """
+        init_col = 0
+        for stg, subenv in self.subenvs.items():
```

Comment:
This is super minor optimization, but is there any reason not to do it only once in init?

Reply:
Ok, I am not repeating this for all of the methods below.